File: modules/expo-pose-detection/android/src/main/java/expo/modules/posedetection/LatestNativeEventScheduler.kt
package expo.modules.posedetection

internal enum class LatestNativeEventType {
  SCHEDULED,
  COALESCED,
  REJECTED,
  EMITTED,
  CANCELLED,
}

internal data class LatestNativeEvent(
  val type: LatestNativeEventType,
  val order: Long,
  val replacedOrder: Long? = null,
  val latestOrder: Long? = null,
)

/**
 * Main-thread event backpressure for native-to-JS pose events.
 *
 * Newer accepted results replace the pending event while one main-thread
 * emission is already scheduled. This keeps event delivery latest-only without
 * building a queue of stale pose payloads under JS/main-thread pressure.
 */
internal class LatestNativeEventScheduler<T>(
  private val getOrder: (T) -> Long,
  private val schedule: (Runnable) -> Unit,
  private val cancelScheduled: (Runnable) -> Unit,
  private val emit: (T) -> Unit,
  private val onEvent: (LatestNativeEvent) -> Unit = {},
) {
  private val lock = Any()
  private var latestValue: T? = null
  private var latestPendingOrder: Long? = null
  private var lastAcceptedOrder: Long? = null
  private var scheduledRunnable: Runnable? = null

  fun submit(value: T): LatestNativeEventType {
    val order = getOrder(value)
    var runnableToSchedule: Runnable? = null
    val eventType: LatestNativeEventType
    val event: LatestNativeEvent

    synchronized(lock) {
      val acceptedOrder = lastAcceptedOrder
      if (acceptedOrder != null && order <= acceptedOrder) {
        eventType = LatestNativeEventType.REJECTED
        event = LatestNativeEvent(
          type = eventType,
          order = order,
          latestOrder = acceptedOrder,
        )
      } else {
        val replacedOrder = latestPendingOrder
        latestValue = value
        latestPendingOrder = order
        lastAcceptedOrder = order

        if (scheduledRunnable != null) {
          eventType = LatestNativeEventType.COALESCED
          event = LatestNativeEvent(
            type = eventType,
            order = order,
            replacedOrder = replacedOrder,
          )
        } else {
          val runnable = Runnable {
            val selected: T?
            val selectedOrder: Long
            synchronized(lock) {
              selected = latestValue
              selectedOrder = latestPendingOrder ?: -1L
              latestValue = null
              latestPendingOrder = null
              scheduledRunnable = null
            }
            if (selected != null) {
              emit(selected)
              onEvent(
                LatestNativeEvent(
                  type = LatestNativeEventType.EMITTED,
                  order = selectedOrder,
                )
              )
            }
          }
          scheduledRunnable = runnable
          runnableToSchedule = runnable
          eventType = LatestNativeEventType.SCHEDULED
          event = LatestNativeEvent(type = eventType, order = order)
        }
      }
    }

    runnableToSchedule?.let(schedule)
    onEvent(event)
    return eventType
  }

  fun cancel() {
    val runnableToCancel: Runnable?
    val order: Long?
    synchronized(lock) {
      runnableToCancel = scheduledRunnable
      order = latestPendingOrder
      latestValue = null
      latestPendingOrder = null
      scheduledRunnable = null
    }
    if (runnableToCancel != null) {
      cancelScheduled(runnableToCancel)
      onEvent(
        LatestNativeEvent(
          type = LatestNativeEventType.CANCELLED,
          order = order ?: -1L,
        )
      )
    }
  }

  fun reset() {
    synchronized(lock) {
      latestValue = null
      latestPendingOrder = null
      lastAcceptedOrder = null
      scheduledRunnable = null
    }
  }
}

File: modules/expo-pose-detection/ios/LatestNativeEventScheduler.swift
import Foundation

enum LatestNativeEventType {
  case scheduled
  case coalesced
  case rejected
  case emitted
  case cancelled
}

struct LatestNativeEventStats {
  let scheduled: Int
  let coalesced: Int
  let rejected: Int
  let emitted: Int
}

/**
 * Applies latest-only backpressure before native pose results reach the main
 * thread. A newer result replaces the one pending delivery, so a busy main
 * thread never builds a queue of stale mask/landmark pairs.
 */
final class LatestNativeEventScheduler<Value> {
  private let getOrder: (Value) -> Int
  private let schedule: (@escaping () -> Void) -> Void
  private let emit: (Value) -> Void
  private let onEvent: (LatestNativeEventType) -> Void
  private let lock = NSLock()

  private var latestValue: Value?
  private var latestPendingOrder: Int?
  private var lastAcceptedOrder: Int?
  private var scheduledGeneration = 0
  private var deliveryScheduled = false
  private var scheduledCount = 0
  private var coalescedCount = 0
  private var rejectedCount = 0
  private var emittedCount = 0

  init(
    getOrder: @escaping (Value) -> Int,
    schedule: @escaping (@escaping () -> Void) -> Void,
    emit: @escaping (Value) -> Void,
    onEvent: @escaping (LatestNativeEventType) -> Void = { _ in }
  ) {
    self.getOrder = getOrder
    self.schedule = schedule
    self.emit = emit
    self.onEvent = onEvent
  }

  @discardableResult
  func submit(_ value: Value) -> LatestNativeEventType {
    let order = getOrder(value)
    var generationToSchedule: Int?
    let event: LatestNativeEventType

    lock.lock()
    if let acceptedOrder = lastAcceptedOrder, order <= acceptedOrder {
      rejectedCount += 1
      event = .rejected
    } else {
      latestValue = value
      latestPendingOrder = order
      lastAcceptedOrder = order
      if deliveryScheduled {
        coalescedCount += 1
        event = .coalesced
      } else {
        deliveryScheduled = true
        scheduledGeneration += 1
        generationToSchedule = scheduledGeneration
        scheduledCount += 1
        event = .scheduled
      }
    }
    lock.unlock()

    if let generationToSchedule {
      schedule { [weak self] in
        self?.drain(generation: generationToSchedule)
      }
    }
    onEvent(event)
    return event
  }

  func cancel() {
    var cancelled = false
    lock.lock()
    if deliveryScheduled || latestValue != nil {
      cancelled = true
    }
    latestValue = nil
    latestPendingOrder = nil
    deliveryScheduled = false
    scheduledGeneration += 1
    lock.unlock()
    if cancelled { onEvent(.cancelled) }
  }

  func reset() {
    lock.lock()
    latestValue = nil
    latestPendingOrder = nil
    lastAcceptedOrder = nil
    deliveryScheduled = false
    scheduledGeneration += 1
    scheduledCount = 0
    coalescedCount = 0
    rejectedCount = 0
    emittedCount = 0
    lock.unlock()
  }

  func stats() -> LatestNativeEventStats {
    lock.lock()
    let value = LatestNativeEventStats(
      scheduled: scheduledCount,
      coalesced: coalescedCount,
      rejected: rejectedCount,
      emitted: emittedCount)
    lock.unlock()
    return value
  }

  private func drain(generation: Int) {
    let selected: Value?
    lock.lock()
    if deliveryScheduled, generation == scheduledGeneration {
      selected = latestValue
      latestValue = nil
      latestPendingOrder = nil
      deliveryScheduled = false
    } else {
      selected = nil
    }
    lock.unlock()

    guard let selected else { return }
    emit(selected)
    lock.lock()
    emittedCount += 1
    lock.unlock()
    onEvent(.emitted)
  }
}
